### PR TITLE
Client listing branch

### PR DESCRIPTION
**Pull request by bbockelm**
This is a new version of PR #1257.

Currently marked as a draft: there's a fundamental issue where the return value from the library is a big string.

Instead, we should be returning a slice of objects (analogous to `readdir` and friends in POSIX) and then the CLI should convert that into strings.  Some implications of this:
- Remove the "directories only" and "files only" option from the library.  We shouldn't add things if they're not truly necessary (can be handled in the post-processing).
- Remove the "JSON" option.

---
From @haoming29: 
The revised PR addressed the above issue as well as the future change in how the director responds to `PROPFIND` request. 

For the result handling, I removed most of the unit tests from the client library for different style flag (dirOnly, json, etc), but I didn't test them in the CLI. This can be a follow up issue.

For `PROPFIND`, the client now recognize director's response with `207` status code and `Pelican-X-Namespace` URL, and the client then reads from the `collections-url` of that header for the collections url. If the director still returns `307`, then the client reads the `Location` header and strip path of the URL as the collection URL instead.

To test this call, try `pelican object ls osdf:///et-gw/PUBLIC` which is a listing enabled origin. Try `-D` for directory only result, `-O` for object only result, `-L` for detail info, and `-j` for JSON-formatted result.

### CODE REVIEW

**Comment by bbockelm** [comment]
@haoming29 - as you revise this branch, please take the future #1394 into account.

That is, at some point after 7.9, we may proxy the PROPFIND request at the director instead of redirect at the origin.  This should not break the 7.9 clients.

**Comment by haoming29** [comment]
One unanswered question I spot in the original PR is about how we should get the `collectionsUrl`. In the current director code, we return 307 for `PROPFIND` response with location being the origin URL who has `listing` enabled. Meanwhile the `X-Pelican-Namespace` header includes a simiar information as `collections-url` field. They differ slightly in defining an origin with listing enabled, but I'm not sure if we want to 1. make the source of information consistent 2. use one way over the other for getting the collection URL

**Comment by haoming29** [comment]
@jhiemstrawisc I have a thought. Let's get this in first than #1400 because this one has the revised `getCollectionsURL` function. I will working on #1400 for the rest of changes but once this is merged, I can rebase and fix up any merge conflict. This way we don't have to be back-and-forth between the two branches.

**Comment by haoming29** [comment]
One thing I'm not sure about is the use of `Run` vs `RunE` in the `cmd/object_ls.go`. I saw all other client CLI uses `Run` and manually call `os.Exit(1)`, but in the listing there are a handful of error checks, and I felt a bit repetitive in calling `log.Errorln` then call `os.Exit(1)`. Do we a preference over the two?

File: cmd/object_ls.go
Comment (haoming29) [reply cmd/object_ls.go:48]: This is what I got when I asked ChatGPT about alternatives to describe directory in an object storage system

> For "Directory":
> * Container: Often used in systems like OpenStack Swift.
> * Namespace: Used to define a distinct, separate storage area.
> * Collection: Implies a grouping of related objects.
> * Vault: Sometimes used to imply secure storage.
> * Library: Can be used to imply an organized collection of objects.
> * Folder: Though less common, it is still used occasionally for familiarity.

Comment (haoming29) [reply cmd/object_ls.go:48]: I personally like "collections" but if we are to define it in a different way in the future then I'd like to avoid using that

Comment (jhiemstrawisc) [reply cmd/object_ls.go:48]: I like that too. Can you change the command and remove references to `dir` in this part of the CLI?

Comment (jhiemstrawisc) [reply cmd/object_ls.go:48]: Although I also like Namespace. For example, if the federation has two namespaces `/foo` and `/foo/bar` that export POSIX where `/foo` has object `ob1` and `dir1`, then what should we define as the expected output `object ls /foo`? I'd argue it should return `bar`, `ob1` and `dir1`, where `/bar` is part of another namespace.

File: client/handle_http.go
Comment (haoming29) [reply client/handle_http.go:2660]: This feels like a deadlock to me. We *can't* guarantee that when the client issues a webDAV request the origin will respond with the same information as it advertises to the director, even if we do a `ls` in the director beforehand, so I think the message we want to tell user here is "this is some unexpected error that you can't really do much besides contacting us"

**Review by jhiemstrawisc — APPROVED** [review]
Changes are all good, feel free to merge as soon as CI tests are green :)

**Comment by haoming29** [comment]
> Changes are all good, feel free to merge as soon as CI tests are green :)

Just fixed an error message and should be good to go once tests are passing

**Pull request merged**
